operator_cxx/contrib/quantization_int8-inl.h

```diff
@@ -0,0 +1,505 @@
+                        const std::vector<TBlob> &aux_states) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    using namespace mxnet_op;
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 4, DType> data_grad;
+    Tensor<xpu, 4, DType> out_data_grad;
+    Tensor<xpu, 4, DType> data;
+    Tensor<xpu, 4, DType> out;
+    if (out_grad[Quantization_int8_enum::kOut].ndim() == 2) {
+      Shape<4> dshape = Shape4(out_grad[Quantization_int8_enum::kOut].shape_[0],
+                               out_grad[Quantization_int8_enum::kOut].shape_[1], 1, 1);
+      data = in_data[Quantization_int8_enum::kData].get_with_shape<xpu, 4, DType>(dshape, s);
+      out = out_data[Quantization_int8_enum::kOut].get_with_shape<xpu, 4, DType>(dshape, s);
+
+      out_data_grad = out_grad[Quantization_int8_enum::kOut].get_with_shape<xpu, 4, DType>(dshape, s);
+      data_grad = in_grad[Quantization_int8_enum::kData].get_with_shape<xpu, 4, DType>(dshape, s);
+    } else {
+      data = in_data[Quantization_int8_enum::kData].get<xpu, 4, DType>(s);
+      out = out_data[Quantization_int8_enum::kOut].get<xpu, 4, DType>(s);
+
+      out_data_grad = out_grad[Quantization_int8_enum::kOut].get<xpu, 4, DType>(s);
+      data_grad = in_grad[Quantization_int8_enum::kData].get<xpu, 4, DType>(s);
+    }
+    // Assign(data_grad, req[Quantization_int8_enum::kOut], out_data_grad);
+    if (param_.grad_mode == std::string("ste") || param_.is_weight) {
+      mshadow::Copy(data_grad, out_data_grad, s);
+    }
+    else if (param_.grad_mode == std::string("clip")) {
+      uint64_t WORKSPACE_LIMIT = 1024 * 1024 * param_.workspace; // the in_data shape is large
+      Tensor<xpu, 1, uint8_t> workspace = ctx.requested[Quantization_int8_enum::kTempSpace]
+        .get_space_typed<xpu, 1, uint8_t>(Shape1(WORKSPACE_LIMIT), s);
+      uint64_t allocated_bytes = 0ULL;
+      Tensor<xpu, 4, DType> clip_condition(reinterpret_cast<DType*>(workspace.dptr_ + allocated_bytes), data.shape_, s);
+      allocated_bytes += clip_condition.shape_.Size() * sizeof(DType);
+      Tensor<xpu, 4, DType> zero_bc(reinterpret_cast<DType*>(workspace.dptr_ + allocated_bytes), data.shape_, s);
+      allocated_bytes += zero_bc.shape_.Size() * sizeof(DType);
+      
+      Tensor<xpu, 1, DType> temp(reinterpret_cast<DType*>(workspace.dptr_ + allocated_bytes), Shape1(1), s);
+      allocated_bytes += temp.shape_.Size() * sizeof(DType);
+      
+      CHECK_LE(allocated_bytes, WORKSPACE_LIMIT) << " in backward";
+
+      temp =  ScalarExp<DType>(0.0f);
+      zero_bc = broadcast_scalar(temp, zero_bc.shape_);
+
+
+      DType threshold = DType(0.0f);
+      Stream<cpu> *s_cpu = ctx.get_stream<cpu>();
+      Tensor<cpu, 1, DType> threshold_tensor(&threshold, Shape1(1), s_cpu);
+      Tensor<xpu, 1, DType> aux = aux_states[Quantization_int8_enum::kMinmax].get<xpu, 1, DType>(s);
+      mshadow::Copy(threshold_tensor, aux, s);
+      threshold = DType(0.02f);
```

Comment:
fix hardcoding.

Reply:
sorry, `threshold=DType(0.02f)` is a mistake. the `clip` mode is not used currently.